### PR TITLE
Support setting CiviForm server environment variables directly

### PR DESCRIPTION
**Pull request by avrittrohwer**
Supports setting CiviForm server environment variables defined in civiform/civiform server/conf/env-var-docs.json without defining them in the variable_definitions.json files in this repo.

This functionality is enabled by the env_var_docs.parser package being available to import.  Support for making that package available is implemented in https://github.com/civiform/civiform-deploy/pull/34/files.  If the package is not importable, this change is a no-op.

TODO:

- [ ] Read env-var-docs.json file from github based on CIVIFORM_VERSION.

Tested:

In my config file, I set:

```
export CIVIFORM_SUPPORTED_LANGUAGES="en-US, am, zh-TW"
```

This resulted in the generated .tfvars file having:

```
civiform_server_environment_variables = {
  "LOGIN_RADIUS_METADATA_URI"=""
  "STAGING_HOSTNAME"="avritt-dev.civiform.dev"
  "BASE_URL"="https://avritt-dev.civiform.dev"
  "APPLICANT_OIDC_LOCALE_ATTRIBUTE"=""
  "APPLICANT_OIDC_MIDDLE_NAME_ATTRIBUTE"=""
  "APPLICANT_OIDC_FIRST_NAME_ATTRIBUTE"="name"
  "AWS_REGION"="us-east-1"
  "APPLICANT_OIDC_LAST_NAME_ATTRIBUTE"=""
  "LOGIN_RADIUS_API_KEY"=""
  "ADFS_ADDITIONAL_SCOPES"="allatclaims"
  "CIVIFORM_APPLICANT_IDP"="generic-oidc"
  "CIVIFORM_SUPPORTED_LANGUAGES.0"="en-US"
  "CIVIFORM_SUPPORTED_LANGUAGES.1"="am"
  "CIVIFORM_SUPPORTED_LANGUAGES.2"="zh-TW"
  "APPLICANT_OIDC_RESPONSE_MODE"="form_post"
  "CIVIFORM_TIME_ZONE_ID"="America/Los_Angeles"
  "APPLICANT_OIDC_ADDITIONAL_SCOPES"=""
  "APPLICANT_OIDC_DISCOVERY_URI"="https://civiform-staging.us.auth0.com/.well-known/openid-configuration"
  "APPLICANT_OIDC_PROVIDER_NAME"="OidcClient"
  "APPLICANT_OIDC_EMAIL_ATTRIBUTE"="email"
  "ADFS_DISCOVERY_URI"="https://civiform-staging.us.auth0.com/.well-known/openid-configuration"
  "APPLICANT_OIDC_RESPONSE_TYPE"="id_token token"
  "FAVICON_URL"="https://civiform.us/favicon.png"
  "AD_GROUPS_ATTRIBUTE_NAME"="group"
  "LOGIN_RADIUS_SAML_APP_NAME"=""
}
```

I deployed and only 3 languages were shown: 
image

Note: directly setting `export CIVIFORM_SUPPORTED_LANGUAGES.0="en-US"` does not work in bash because `.` are not allowed in bash variable names: https://unix.stackexchange.com/a/93533

### CODE REVIEW

**Comment by bion** [comment]
> * [ ]  Read env-var-docs.json file from github based on CIVIFORM_VERSION.

@avrittrohwer are you planning to do this TODO in this PR?

**Comment by avrittrohwer** [comment]
> @avrittrohwer are you planning to do this TODO in this PR?

Correct.  I'm hoping to get feedback on everything else besides that part however

**Review by bion — COMMENTED** [review]
Overall looks pretty good, will do another round of review when the TODOs are resolved

File: cloud/shared/bin/lib/config_loader_test.py
Comment (jessbumblebee) [inline cloud/shared/bin/lib/config_loader_test.py:8]: Please add some tests for the code new paths that rely on the server variable definitions, both for the code paths where there are two paths now (e.g. validation) and for new methods (e.g. _load_civiform_server_env_vars()). There may be other, I only picked two examples.

File: cloud/shared/bin/lib/variable_definition_loader.py
Comment (jessbumblebee) [inline cloud/shared/bin/lib/variable_definition_loader.py:13]: Have you tested if a RuntimeError is actually triggered for a duplicate? I was under the impression that json.load removes duplicates in favour of the last one added.

Comment (jessbumblebee) [reply cloud/shared/bin/lib/variable_definition_loader.py:13]: Ps: Please add a test for the duplication check if it is working as intended.

File: cloud/shared/bin/lib/write_tfvars.py
Comment (jessbumblebee) [inline cloud/shared/bin/lib/write_tfvars.py:27]: nit: it looks like the continue isn't needed here because the step is always finished.

**Review by jessbumblebee — COMMENTED** [review]
Generally looks good.

There is one thread of thought that I came across multiple times in this change: There are 3 types of variables: locally used variables, variables that are passed through to terraform and used by infrastructure (not in application.conf) and variables that are passed through and used by the server (present in applicaiton.conf). This change introduces auto generation of the last category, but also keeps the old implementation for all categories. 

What crossed my mind is that it would be good get some clarity about what the desired end state is, document that and make it accessible from the code (or use existing docs if they exist).
A few questions for inspiration: Will duplications of vars be removed (vars that are currently in both variable definitions:infra and server)? Will the multiple variable definitions be kept and what will each of them cover? Besides server variables, do we want to distinguish between infrastructure variables that are locally used or passed to terraform? Do we still need the tfvars field in the future or will the definition file we generate the vars from make it clear if it is a tfvar? How does all of this fit together with the various variables.tf files?

File: cloud/aws/templates/aws_oidc/app.tf
Comment (jessbumblebee) [inline cloud/aws/templates/aws_oidc/app.tf:134]: What's the plan for Azure?

File: .github/workflows/aws-oidc-deploy-test.yaml
Comment (avrittrohwer) [reply .github/workflows/aws-oidc-deploy-test.yaml:19]: Because we no-longer source the config file then read variables from the environment, I need to `echo 'export USE_LOCAL_BACKEND=true' >> civiform_config.sh` in this variable instead.

**Comment by jessbumblebee** [comment]
Hi Bion, there are still a few formatting issues in the presubmit, but  the code is pretty much complete. I'll fix the formatting tomorrow in case you want to have a look already.

**Comment by jessbumblebee** [comment]
Note: the code for enabling the feature is under development in a separate PR as it can be done independently and the PR was growing. Steps for completing the auto generation including all paths and not leaving legacy are here:

https://docs.google.com/document/d/1sM161Y8jZRw4UhZai9c_XjylN7y_yr9yONO914byYko/edit#heading=h.oz618xdedonx

File: cloud/shared/bin/lib/config_loader.py
Comment (bion) [inline cloud/shared/bin/lib/config_loader.py:76]: **nitpick:** 
```suggestion
        environment variables.
```

File: cloud/shared/bin/lib/config_loader.py
Comment (bion) [inline cloud/shared/bin/lib/config_loader.py:62]: **nitpick:** add return type hint?

File: cloud/shared/bin/lib/config_loader.py
Comment (bion) [inline cloud/shared/bin/lib/config_loader.py:311]: **nitpick:** 
```suggestion
            # take precedence over infra variables.
```

File: cloud/shared/bin/lib/config_loader.py
Comment (bion) [inline cloud/shared/bin/lib/config_loader.py:310]: **question (non-blocking):** when would a variable end up in both?

File: cloud/shared/bin/lib/config_loader.py
Comment (bion) [inline cloud/shared/bin/lib/config_loader.py:297]: **nitpick:** iirc empty collections are falsy in python
```suggestion
        if civiform_server_env_var_definitions:
```

**Review by bion — CHANGES_REQUESTED** [review]

File: cloud/shared/bin/lib/config_loader.py
Comment (jessbumblebee) [reply cloud/shared/bin/lib/config_loader.py]: In the future we should remove the variable definitions and use an env-var-docs.json type file instead (discuassed with Bion in my handover today)

File: cloud/shared/bin/lib/config_loader.py
Comment (jessbumblebee) [reply cloud/shared/bin/lib/config_loader.py]: Outdated

File: cloud/shared/bin/lib/env-var-docs.json
Comment (jessbumblebee) [reply cloud/shared/bin/lib/env-var-docs.json:1]: Removed

**Review by bion — APPROVED** [review]
Approved but please do not merge until https://github.com/civiform/civiform/issues/4697 is resolved

**Comment by rockycodes** [comment]
@bion re-requesting a last review from you before merging. a couple of changes:
- merged main into this branch and resolved merge conflicts in `app.tf` due to flags that have been added since this PR went up
- removed references to `get_config_variables` which is a method Avritt deprecated

I tested this out by deploying locally both with these changes turned on (which we can do now since [this](https://github.com/civiform/cloud-deploy-infra/pull/147) is working) and turned off (the default). Both paths work now with no errors! I tested enabling service area validation and languages (both index-list type vars) and they work with these changes 🎉 Excited to get this up and running.

**Review by bion — APPROVED** [review]

**Pull request merged**
